core/src/main/java/org/apache/accumulo/fate/Fate.java

```diff
@@ -300,6 +335,55 @@ public TStatus waitForCompletion(long tid) {
     return store.waitForStatusChange(tid, FINISHED_STATES);
   }
 
+  /**
+   * Attempts to cancel a running Fate transaction
+   *
+   * @param tid
+   *          transaction id
+   * @return true if transaction transitioned to a failed state or already in a completed state,
+   *         false otherwise
+   */
+  public boolean cancel(long tid) {
+    String tidStr = Long.toHexString(tid);
+    for (int retries = 0; retries < 5; retries++) {
+      if (store.tryReserve(tid)) {
+        try {
+          TStatus status = store.getStatus(tid);
+          if (status == NEW || status == SUBMITTED || status == IN_PROGRESS) {
+            store.setProperty(tid, EXCEPTION_PROP, new TApplicationException(
+                TApplicationException.INTERNAL_ERROR, "Fate transaction cancelled by user"));
+            store.setStatus(tid, FAILED_IN_PROGRESS);
+            log.info(
+                "Updated status for Repo {} to FAILED_IN_PROGRESS because it was cancelled by user",
+                tidStr);
+            return true;
+          } else {
+            log.info("Repo {} cancelled by user but already in finished state", tidStr);
+            return true;
+          }
+        } finally {
+          store.unreserve(tid, 0);
+        }
+      } else {
+        // It's possible that the FateOp is being run by the TransactionRunner
+        Thread t = null;
+        synchronized (RUNNING_TRANSACTIONS) {
+          t = RUNNING_TRANSACTIONS.get(tid);
+        }
+        if (t != null) {
+          t.interrupt();
```

Comment:
There is a race condition here.  The same thread runs many different FATE tx steps.  By the time this code executes, the thread could be running a completely different FATE tx,  so it could interrupt the wrong FATE tx.

Reply:
What if I moved lines 373 to 381 in the synchronized block? Then, it should be guaranteed that if the transaction is in RUNNING_TRANSACTIONS, that the Thread running it has not moved on to something else.

Reply:
I might also need to move the RUNNING_TRANSACTIONS.remove(tid) up a little and tighten that up.

Reply:
I re-worked the synchronization in [6d22ea7](https://github.com/apache/accumulo/pull/2467/commits/6d22ea7708b786a9b9cd2aa4469e332e29e57476)

Reply:
> I re-worked the synchronization in [6d22ea7](https://github.com/apache/accumulo/pull/2467/commits/6d22ea7708b786a9b9cd2aa4469e332e29e57476)

That may be a good solution to the problem.  I suggested some tweaks to the approach.